vignettes/elpd.Rmd

```diff
@@ -0,0 +1,213 @@
+
+```{r gen-test}
+gen_test <- gqs(stanmodel, draws = as.matrix(fit), data= data_test)
+log_lik_test <- extract_log_lik(gen_test)
+```
+
+## Computing holdout elpd:
+
+Now we evaluate the predictive performance of the model on the test data using `elpd()`.
+
+```{r elpd-holdout}
+(elpd_holdout <- elpd(log_lik_test))
+```
+
+When one wants to compare different models, the function `loo_compare()` can be used to assess the difference in performance.
+
+# K-fold cross validation
+
+For this approach the data is divided into folds, and each time one fold is tested while the rest of the data is used to fit the model (see Vehtari et al., 2017).
+
+## Splitting the data in folds
+
+We use the data that is already pre-processed and we divide it in 10 random folds using `kfold_split_random`
+
+```{r prepare-folds, message=FALSE}
+# Prepare data
+roaches$fold <- kfold_split_random(K = 10, N = nrow(roaches))
+```
+
+
+## Fitting and extracting the likelihood of each fold
+
+We now loop over the 10 folds: Each time, we fit the model to all the observations except the ones belonging to the fold, but we extract the log-likelihood from the fold. The output of this loop is a list of log-likelihood matrices.
+
+```{r}
+log_lik_list <- lapply(1:10, function(k){
+  data_train <- list(y = roaches$y[roaches$fold != k],
+                   x = as.matrix(roaches[roaches$fold != k,
+                                         c("roach1", "treatment", "senior")]),
+                   N = nrow(roaches[roaches$fold != k,]),
+                   K = 3,
+                   offset = roaches$offset[roaches$fold != k],
+                   beta_prior_scale = 2.5,
+                   alpha_prior_scale = 5.0
+                   )
+  data_test <- list(y = roaches$y[roaches$fold == k],
+                   x = as.matrix(roaches[roaches$fold == k,
+                                         c("roach1", "treatment", "senior")]),
+                   N = nrow(roaches[roaches$fold == k,]),
+                   K = 3,
+                   offset = roaches$offset[roaches$fold == k],
+                   beta_prior_scale = 2.5,
+                   alpha_prior_scale = 5.0
+                   )
+  fit <- sampling(stanmodel, data = data_train, seed = seed, refresh = 0)
+  gen_test <- gqs(stanmodel, draws = as.matrix(fit), data= data_test)
+  log_lik_test <- extract_log_lik(gen_test)
+  log_lik_test
+})
+```
+
+## Computing k-fold elpd:
+
+Now we evaluate the predictive performance of the model on the 10 folds using `elpd()`.
+
+```{r elpd-kfold}
+# First we need to convert the list of matrices into a long matrix:
+log_lik_kfold <- do.call(cbind, log_lik_list) 
+(elpd_kfold <- elpd(log_lik_kfold))
+```
+
+If one wants to compare several models (with `loo_compare`), one should use the same folds for all the different models. (Notice that now the likelihood values in `log_lik_kfold` are not in the same order as the observations of `roaches`.) 
```

Comment:
"(Notice that now the likelihood values in `log_lik_kfold` are not in the same order as the observations of `roaches`.) "
Why not? I think they should be in the same order. You can check rstanarm::kfold for how to arrange the values back in the original order.

Reply:
sorry, do you mean that what I said is wrong or that it's right but I should implement it in a way that the values are back in the right order?

Reply:
Please, implement it in a way that the values are back in the right order, so that then it's easy to see outliers and compare to loo values.